### PR TITLE
Support aws S3 as file storage

### LINKED ISSUE #34: Support Amazon S3 as a file store
**What would you like to be added**:

**Why is this needed**:

### PR DESCRIPTION
**Pull request by khanhtc1202**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes**:

Fixes #34 

**Does this PR introduce a user-facing change?**:

```release-note
User can config s3 as file storage for pipecd
```

### CODE REVIEW

File: pkg/config/control_plane.go [hunk 1]
```diff
@@ -314,6 +314,14 @@ type FileStoreGCSConfig struct {
 }
 
 type FileStoreS3Config struct {
+	// The bucket name to store artifacts and logs in the pipe.
```
Comment (nghialv) [inline pkg/config/control_plane.go:317]: typo: `pipe` -> `piped`

File: pkg/filestore/s3/s3.go [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+		func(page *s3.ListObjectsV2Output, lastPage bool) bool {
+			for _, obj := range page.Contents {
+				objects = append(objects, filestore.Object{
+					Path:    aws.StringValue(obj.Key),
+					Size:    aws.Int64Value(obj.Size),
+					Content: []byte{},
+				})
+			}
+			return *page.IsTruncated
+		})
```
Comment (nghialv) [inline pkg/filestore/s3/s3.go:168]: nit:

```
   },
)
```

**Comment by nghialv** [comment]
Nice. I have just left some nits.

File: pkg/filestore/s3/s3.go (hunk 2 above)
Comment (khanhtc1202) [reply pkg/filestore/s3/s3.go]: thank you, addressed on https://github.com/pipe-cd/pipe/pull/1485/commits/8dea680e82b2a373886f4c69bebc9c9fee066b03 🙏

File: pkg/config/control_plane.go (hunk 1 above)
Comment (khanhtc1202) [reply pkg/config/control_plane.go]: thank you, addressed on https://github.com/pipe-cd/pipe/pull/1485/commits/8dea680e82b2a373886f4c69bebc9c9fee066b03 🙏

**Comment by nghialv** [comment]
Thank you.
/lgtm

File: pkg/filestore/s3/s3.go [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+}
+
+type Option func(*Store)
+
+func WithLogger(logger *zap.Logger) Option {
+	return func(s *Store) {
+		s.logger = logger.Named("s3")
+	}
+}
+
```
Comment (nakabonne) [inline pkg/filestore/s3/s3.go:50]: nit: The credential file is optional right? So I feel like it's better to set the path to the credentials file via `Option` like:

```go
func WithCredentialsFile(path string) Option {
	return func(s *Store) {
		s.credentialsFile = path
	}
}
```

Just want to unify with other parts.

File: pkg/filestore/s3/s3.go [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+func (s *Store) GetObject(ctx context.Context, path string) (object filestore.Object, err error) {
+	object.Path = path
+	rc, err := s.NewReader(ctx, path)
+	if err != nil {
+		return
+	}
+	content, err := ioutil.ReadAll(rc)
+	if err != nil {
+		rc.Close()
```
Comment (nakabonne) [inline pkg/filestore/s3/s3.go:123]: I feel just enclosing defer statement is better. I'm aware that GCS's one is the same. But never mind, I'd say that should be so as well.

```go
	defer func() {
		if err := rc.Close(); err != nil {
			s.logger.Error("failed to close...", zap.Error(err))
		}
	}()
```

File: pkg/filestore/s3/s3.go (hunk 3 above)
Comment (khanhtc1202) [reply pkg/filestore/s3/s3.go:52]: addressed on https://github.com/pipe-cd/pipe/pull/1485/commits/065211a7f92aefe59cdf7fba4255e1a3a67bca6e 🙏

File: pkg/filestore/s3/s3.go (hunk 4 above)
Comment (khanhtc1202) [reply pkg/filestore/s3/s3.go]: Addressed on https://github.com/pipe-cd/pipe/pull/1485/commits/95573edcdacb1d3a5512c7cb0c5952d9eb9ca0a6 🙏 PTAL 👀

File: pkg/filestore/s3/s3.go [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+
+	content, err := ioutil.ReadAll(rc)
+	defer func() {
+		if e := rc.Close(); e != nil {
+			if err == nil {
+				err = fmt.Errorf("unable to close read object stream: %w", e)
+			} else {
+				err = fmt.Errorf("%w and unable to close read object stream: %s", err, e.Error())
+			}
+		}
```
Comment (nakabonne) [inline pkg/filestore/s3/s3.go:138]: I'd say upon it calls `s.NewReader(ctx, path)`, should define this part. Besides, I think we need to check only error emitted by `rc.Close()` because `err` is being handled appropriately.

Comment (khanhtc1202) [reply pkg/filestore/s3/s3.go:138]: That exactly what I want your point 👀 I don't have a strong opinion about this, but is it okay to just log the error given by `rc.Close()` with `s.logger` 👀

Comment (nakabonne) [reply pkg/filestore/s3/s3.go:138]: Whether ignoring an error from Close() is good or bad. I think this topic is highly depending on its context, but in this case, logging could be enough. That get ends with success.

Comment (nakabonne) [reply pkg/filestore/s3/s3.go:138]: But I also have a non-strong opinion. I'd be happy to hear your thoughts.

Comment (khanhtc1202) [reply pkg/filestore/s3/s3.go:138]: Currently, in case of minio storage, we do ignore the handling (just show it on the log)
https://github.com/pipe-cd/pipe/blob/965abdc59242c57a5051ac3817d2bd583a5a28d4/pkg/filestore/minio/minio.go#L125
but in case of GCS we return it for further handling (maybe it's just be shown to the log somewhere either 😂 )
https://github.com/pipe-cd/pipe/blob/9ce7102afd38da0046e595f4dacc61aec9562e76/pkg/filestore/gcs/gcs.go#L108-L111
In short, we do not have a unified way to handle error from `rc.Close()` currently 😅 I think I will go as your recommended way to handle (logging with `s.logger` is enough). Let's me make separated PR to make the current GCS handling `rc.Close()` error to unify with other storages.

Comment (nakabonne) [reply pkg/filestore/s3/s3.go:138]: Fine, I feel exactly the same way!

Comment (khanhtc1202) [reply pkg/filestore/s3/s3.go:138]: okay let's me fix this part of s3 storage first 🤝

Comment (khanhtc1202) [reply pkg/filestore/s3/s3.go]: Addressed on https://github.com/pipe-cd/pipe/pull/1485/commits/69df4dcbd640c91d45d6313cdce0e8cb82f29db4 🙏

**Comment by nakabonne** [comment]
Thank you 🤝 
/approve

**Comment by khanhtc1202** [comment]
> Thank you 🤝
> /approve

🤝
